File: javalin/javalin-java-HelloWorldController/src/main/java/com/varmateo/controller/HelloWorldController.java
/**
 *
 */
package com.varmateo.controller;

import io.javalin.Context;


public final class HelloWorldController {


    public static final String PATH_HELLO = "/hello";
    public static final String PATH_GREETINGS = "/greetings";
    public static final String PATH_GREET_MESSAGE = "/greet-message";

    private static final String PARAM_INPUT = "input";
    private static final String PARAM_NAME = "name";


    public HelloWorldController() {
        // Nothing to do.
    }


    public void hello(final Context context) {
        context.result("Hello, world!");
    }


    public void greetings(final Context context) {

        String name = context.queryParam(PARAM_NAME);

        if ( name != null ) {
            String response = String.format("Hello, %s", name);
            context.result(response);
        } else {
            context.status(400); // Bad request
        }
    }


    public void greetMessage(final Context context) {

        String input = context.queryParam(PARAM_INPUT);

        if ( input != null ) {
            GreetMessageResponse response = new GreetMessageResponse(
                    input,
                    String.format("Hello, %s", capitalize(input)));
            context.json(response);
        } else {
            context.status(400); // Bad request
        }
    }


    private String capitalize(final String s) {

        if ( s.length() == 0 ) {
            return s;
        } else {
            return new StringBuilder()
                    .append(Character.toUpperCase(s.charAt(0)))
                    .append(s.substring(1).toLowerCase())
                    .toString();
        }
    }

}

File: vertx/vertx-kotlin-HelloWorldController/src/main/kotlin/com/varmateo/controller/HelloWorldController.kt
package com.varmateo.controller

import io.vertx.core.http.HttpMethod
import io.vertx.core.http.HttpServerResponse
import io.vertx.core.json.Json
import io.vertx.ext.web.Router
import io.vertx.ext.web.RoutingContext

import com.varmateo.controller.GreetMessageResponse


private const val PARAM_INPUT = "input"
private const val PARAM_NAME = "name"


class HelloWorldController {

    val hello = get("/hello", ::doHello)
    val greetings = get("/greetings", ::doGreetings)
    val greetMessage = get("/greet-message", ::doGreetMessage)


    private fun doHello(context: RoutingContext): Unit {
        context.response().end("Hello, world!")
    }


    fun doGreetings(context: RoutingContext): Unit {
        val name: String? = context.request().getParam(PARAM_NAME)
        val response: HttpServerResponse = context.response()

        if (name != null) {
            response
                    .setChunked(true)
                    .write("Hello, ${name}")
        } else {
            response.setStatusCode(400)
        }

        response.end()
    }


    fun doGreetMessage(context: RoutingContext): Unit {
        var input: String? = context.request().getParam(PARAM_INPUT)
        var response: HttpServerResponse = context.response()

        if (input != null) {
            var result = GreetMessageResponse(
                    input = input,
                    output = "Hello, ${input.capitalize()}!")

            response
                    .setChunked(true)
                    .putHeader("content-type", "application/json; charset=utf-8")
                    .write(Json.encodePrettily(result))
        } else {
            response.setStatusCode(400)
        }

        response.end()
    }
}


data class RouteDescription(
        val method: HttpMethod,
        val path: String,
        val handler: (RoutingContext) -> Unit)


fun Router.addRoute(routeDescription: RouteDescription): Router {

    this.route(routeDescription.path)
            .method(routeDescription.method)
            .handler(routeDescription.handler)

    return this
}


private fun get(
        path: String,
        handler: (RoutingContext) -> Unit)
        : RouteDescription = RouteDescription(HttpMethod.GET, path, handler)
